src/main/java/net/dv8tion/jda/api/utils/messages/MessageCreateBuilder.java

```diff
@@ -0,0 +1,288 @@
+        return this;
+    }
+
+    @Nonnull
+    @Override
+    public MessageCreateBuilder addComponents(@Nonnull Collection<? extends LayoutComponent> components)
+    {
+        Checks.noneNull(components, "ComponentLayouts");
+        for (LayoutComponent layout : components)
+            Checks.check(layout.isMessageCompatible(), "Provided component layout is invalid for messages!");
+        Checks.check(
+                this.components.size() + components.size() <= Message.MAX_COMPONENT_COUNT,
+            "Cannot add more than %d component layouts", Message.MAX_COMPONENT_COUNT
+        );
+        this.components.addAll(components);
+        return this;
+    }
+
+    @Nonnull
+    @Override
+    public MessageCreateBuilder setFiles(@Nullable Collection<? extends FileUpload> files)
+    {
+        if (files != null)
+            Checks.noneNull(files, "Files");
+        this.files.clear();
+        if (files != null)
+            this.files.addAll(files);
+        return this;
+    }
+
+    @Nonnull
+    @Override
+    public List<? extends FileUpload> getAttachments()
+    {
+        return Collections.unmodifiableList(files);
+    }
+
+    @Nonnull
+    @Override
+    public MessageCreateBuilder addFiles(@Nonnull Collection<? extends FileUpload> files)
+    {
+        Checks.noneNull(files, "Files");
+        this.files.addAll(files);
+        return this;
+    }
+
+    @Nonnull
+    @Override
+    public MessageCreateBuilder setTTS(boolean tts)
+    {
+        this.tts = tts;
+        return this;
+    }
+
+    @Nonnull
+    @Override
+    public MessageCreateBuilder setSuppressEmbeds(boolean suppress)
+    {
+        int flag = MessageFlag.EMBEDS_SUPPRESSED.getValue();
+        if (suppress)
+            this.flags |= flag;
+        else
+            this.flags &= ~flag;
+        return this;
+    }
+
+    @Override
+    public boolean isSuppressEmbeds()
+    {
+        return (this.flags & MessageFlag.EMBEDS_SUPPRESSED.getValue()) != 0;
+    }
+
+    @Override
+    public boolean isEmpty()
+    {
+        return Helpers.isBlank(content) && embeds.isEmpty() && files.isEmpty();
+    }
+
+    @Override
+    public boolean isValid()
+    {
+        return !isEmpty() && embeds.size() <= Message.MAX_EMBED_COUNT
+                          && components.size() <= Message.MAX_COMPONENT_COUNT
+                          && Helpers.codePointLength(content) <= Message.MAX_CONTENT_LENGTH;
+    }
+
+    @Nonnull
+    public MessageCreateData build()
+    {
+        // Copy to prevent modifying data after building
+        String content = this.content.toString().trim();
+        List<MessageEmbed> embeds = new ArrayList<>(this.embeds);
+        List<FileUpload> files = new ArrayList<>(this.files);
+        List<LayoutComponent> components = new ArrayList<>(this.components);
+        AllowedMentionsImpl allowedMentions = this.allowedMentions.copy();
+
+        if (content.isEmpty() && embeds.isEmpty() && files.isEmpty())
```

Comment:
Is there a reason this doesn't call `isEmpty`?

Reply:
Thread safety, since we copy the data out before this.